File: 833-bus-routes/bus-routes.js
/**
 * @param {number[][]} routes
 * @param {number} source
 * @param {number} target
 * @return {number}
 */
var numBusesToDestination = function(routes, source, target) {
    let q = new Queue();
    let set = new Set();
    let map = new Map();
    let vis = new Set();
    
    for(let i=0; i<routes.length; i++) {
        for(let stop of routes[i]) {
            if(!map.has(stop)) map.set(stop, []);
            map.get(stop).push(i);
        }
    }

    q.enqueue(source);
    set.add(source);
    let ans = 0;

    while(q.size() > 0) {
        let size = q.size();
        //console.log(q);
        for(let s=0; s<size; s++) {
            let curr = q.dequeue();
            if(curr === target) return ans;
            for(let bus of map.get(curr)) {
                if(vis.has(bus)) continue;
                vis.add(bus);
                for(let stop of routes[bus]) {
                    if(!set.has(stop)) {
                        set.add(stop);
                        q.enqueue(stop);
                    }
                }
            }
        }
        ans++;
    }

    return -1;
};

File: 833-bus-routes/bus-routes.java
class Solution {
    public int numBusesToDestination(int[][] routes, int source, int target) {
        if(source == target) return 0;
        Map<Integer, List<Integer>> map = new HashMap<>();
        for(int i=0; i<routes.length; i++) {
            for(int stop : routes[i]) {
                map.computeIfAbsent(stop, a -> new ArrayList<>()).add(i);
            }
        }
        Set<Integer> buses = new HashSet<>();
        Set<Integer> stops = new HashSet<>();
        Queue<Integer> q = new LinkedList<>();
        // for(int bus : map.getOrDefault(source)) {
        //     buses.add(bus);
        //     for(int stop : routes[bus]) {
        //         if(stops.add(stop)) {
        //             q.add(stop);
        //         }
        //     }
        // }
        q.add(source);
        int ans = 0;
        while(!q.isEmpty()) {
            int size = q.size();
            while(size-- > 0) {
                int curr = q.poll();
                if(curr == target) return ans;
                for(int bus : map.getOrDefault(curr, new ArrayList<>())) {
                    if(buses.add(bus)) {
                        for(int stop : routes[bus]) {
                            if(stops.add(stop)) {
                                q.add(stop);
                            }
                        }
                    }
                }
            }
            ans++;
        }
        return -1;
    }
}